sips/sip-6.md

```diff
@@ -0,0 +1,353 @@
+    stake_activation_epoch: u64,
+    ctx: &mut TxContext
+): StakedSui {
+    let sui_amount = balance::value(&stake);
+    assert!(!is_inactive(pool), EDelegationToInactivePool);
+    assert!(sui_amount > 0, EDelegationOfZeroSui);
+
+    pool.pending_stake = pool.pending_stake + sui_amount;
+
+    StakedSui {
+        id: object::new(ctx),
+        pool_id: object::id(pool),
+        stake_activation_epoch,
+        principal: stake,
+    }
+}
+```
+
+The above example follows the notation set by an implementation of the `request_withdraw_stake` flow that returns `Balance<SUI>` which, at the time of writing, has just been included into Sui version `testnet-v1.3.0`.<sup>[3](https://github.com/MystenLabs/sui/pull/12092)</sup>
+
+### iii. `sui_system` getters
+
+#### iiia. `calculate_rewards` function
+
+**sui_system.** A top level `public` function, `calculate_rewards`, will be added as a wrapper around the flow that will call `staking_pool::calculate_rewards`. This function will allow other contracts to calculate the rewards that a `StakedSui` object is entitled to.
+
+```Rust
+/// Given the `staked_sui` receipt calculate the current rewards (in terms of SUI) for it.
+public fun calculate_rewards(wrapper: &mut SuiSystemState, staked_sui: &StakedSui): u64 {
+    let self = load_system_state(wrapper);
+
+    sui_system_state_inner::calculate_rewards(self, staked_sui)
+}
+```
+
+**sui_system_state_inner.** Following the design of other `sui_system` functions, a corresponding `calculate_rewards` function will be added to `sui_system_state_inner.move`.
+
+```Rust
+/// Given the `staked_sui` receipt calculate the current rewards (in terms of SUI) for it.
+public(friend) fun calculate_rewards(self: &SuiSystemStateInnerV2, staked_sui: &StakedSui): u64 {
+    let validators = &self.validators;
+    let epoch = epoch(self);
+
+    validator_set::calculate_rewards(validators, staked_sui, epoch)
+}
+```
+
+**validator_set.** Similarly, a corresponding `calculate_rewards` function will be added to `validator_set.move`.
+
+```Rust
+/// Given the `staked_sui` receipt calculate the current rewards (in terms of SUI) for it.
+public(friend) fun calculate_rewards(self: &ValidatorSet, staked_sui: &StakedSui, epoch: u64): u64 {
+    let staking_pool_id = pool_id(staked_sui);
+    let mapping = &self.staking_pool_mappings;
+
+    // If there is no such staking pool id in the staking pool id -> validator mapping
+    // return zero.
+    if (!table::contains(mapping, staking_pool_id)) {
+        return 0
+    };
+
+    let validator_address = *table::borrow(mapping, staking_pool_id);
+    // There are no rewards if validator is not active.
+    if (!is_active_validator_by_sui_address(self, validator_address)) {
+        return 0
+    };
+
+    let validator = get_active_validator_ref(self, validator_address);
+    let staking_pool = validator::get_staking_pool_ref(validator);
+
+    staking_pool::calculate_rewards(staking_pool, staked_sui, epoch)
```

Comment:
Hello. Looks good. But I am against using those functions of `staking_pool` module in `validator_set.move`
In original design, functions of `staking_pool` module are designed to be accessible in `validator.move`.
So this section can be amended as following
in `validator_set.move`
```rust
public(friend) fun calculate_rewards(self: &ValidatorSet, staked_sui: &StakedSui, epoch: u64): u64 {
...
   let validator_ref = get_active_validator_ref(self, validator_address);
   validator::calculate_rewards(validator_ref, staked_sui, epoch);
...
}
```
in `validator.move`
```rust
public(friend) fun calculate_rewards(self: &Validator, staked_sui: &StakedSui, epoch: u64): u64 {
   staking_pool::calculate_rewards(&self.staking_pool, staked_sui, epoch)
}
```